src/scipp/core/variable.py

```diff
@@ -287,14 +290,44 @@ def array(*,
                          dtype=dtype)
 
 
+# Assumes that all arguments are Variable or None.
+def _ensure_same_unit(*, unit, args: dict):
+    if unit == default_unit:
+        units = {key: val.unit for key, val in args.items() if val is not None}
+        if len(set(units.values())) != 1:
+            raise _cpp.UnitError(
+                f'All units of the following arguments must be equal: {units}. '
+                'You can specify a unit explicitly with the `unit` argument.')
+        unit = next(iter(units.values()))
+    return {
+        key: _cpp.to_unit(val, unit, copy=False).value if val is not None else None
+        for key, val in args.items()
+    }, unit
+
+
+# Process arguments of arange, linspace, etc and return them as plain numbers or None.
+def _normalize_range_args(*, unit, **kwargs):
+    is_var = {
+        key: isinstance(val, _cpp.Variable)
+        for key, val in kwargs.items() if val is not None
+    }
+    if any(is_var.values()):
+        if not all(is_var.values()):
+            arg_types = {key: type(val) for key, val in kwargs.items()}
+            raise TypeError('Either all of the following arguments or none have to '
+                            f'be variables: {arg_types}')
+        return _ensure_same_unit(unit=unit, args=kwargs)
+    return kwargs, unit
```

Comment:
Missing a check that the variables to not have variances?

Reply:
Yes